### PR TITLE
Feat: #41 Add Sentry logging

### LINKED ISSUE #41: Add Sentry logging
* Could reuse [https://github.com/web3-storage/web3.storage/blob/9fac681b121f2902f0570c15e7e6b4bf16521d3c/packages/api/src/utils/logs.js] from web3.storage

### PR DESCRIPTION
**Pull request by joshghent**
Resolves #41

### CODE REVIEW

File: packages/api/buildworkermodule.mjs [hunk 1]
```diff
@@ -1,23 +1,69 @@
...
+
+      await cli.releases.new(sentryRelease);
+      await cli.releases.setCommits(sentryRelease, {
+        auto: true,
+        ignoreEmpty: true,
+        ignoreMissing: true,
+      });
+      await cli.releases.uploadSourceMaps(sentryRelease, {
+        // validate: true,
+        include: [path.join(__dirname, "..", "dist")],
```
Comment (adamalton) [inline packages/api/buildworkermodule.mjs:58]: From doing a comparison to the same code in web3.storage, by my calculation the `'..'` should be omitted here. This script lives in `packages/api/`, whereas the script which contains the exact same line of code in web3.storage lives in `packages/api/scripts/`. And `__dirname` is defined in the same way in both.

File: packages/api/src/utils/logs.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+   */
+  async end (response: Response) {
+    if (this._finished) {
+      throw new Error(
+        `end() has already been called on this Logging instance.
+        You must make a new instance per request.`
+      )
+    }
+    this._finished = true
+    if (this.opts?.debug != null) {
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:141]: Is this comparison deliberate? `opts.debug` is a boolean, and both `true != null` and `false != null` will both be true.

File: packages/api/src/utils/logs.ts [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+      }
+      if (this.opts?.debug != null) {
+        console[level](`[${dt}]`, message)
+      }
+    } else {
+      log = {
+        ...log,
+        message
+      }
+      if (this.opts?.debug != null) {
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:212]: Ditto re. null comparison.

File: packages/api/src/utils/logs.ts [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    }
+    return timeObj
+  }
+
+  _timersString () {
+    const result = []
+    for (const key of this._timesOrder) {
+      const { name, duration, description }: { name: string, duration: string, description: string } = this._times.get(key)
+      result.push(
+        description === ''
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:302]: This comparison is the wrong way round, if I'm not mistaken?
```suggestion
        description !== ''
```

File: packages/api/src/utils/logs.ts [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+  }
+
+  /**
+   * End the timer of the given name.
+   *
+   * @param {string} name
+   */
+  timeEnd (name: string) {
+    const timeObj = this._times.get(name)
+    if (timeObj == null) {
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:277]: Threequalise for fun and profit?

```suggestion
    if (timeObj === undefined) {
```

File: packages/api/src/utils/logs.ts [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+  }
+
+  /**
+   * Start a timer to be logged under the given name.
+   *
+   * @param {string} name
+   * @param {string} [description]
+   */
+  time (name: string, description?: string) {
+    if (this._times.get(name) != null) {
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:259]: Threequalise?

```suggestion
    if (this._times.get(name) !== undefined) {
```

**Review by adamalton — COMMENTED** [review]
Aside from my comments and the small tweak I've pushed, this overall looks good.

File: packages/api/buildworkermodule.mjs (hunk 1 above)
Comment (joshghent) [reply packages/api/buildworkermodule.mjs:58]: Yup, you're right. Corrected this. Good spot.

File: packages/api/src/utils/logs.ts (hunk 6 above)
Comment (joshghent) [reply packages/api/src/utils/logs.ts]: Good spot, thank you :)

File: packages/api/src/utils/logs.ts (hunk 5 above)
Comment (joshghent) [reply packages/api/src/utils/logs.ts]: Updated, good spot

File: packages/api/src/utils/logs.ts (hunk 3 above)
Comment (joshghent) [reply packages/api/src/utils/logs.ts]: Updated to `?? false` as per https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md

File: packages/api/src/utils/logs.ts (hunk 2 above)
Comment (joshghent) [reply packages/api/src/utils/logs.ts]: Good spot, this was an incorrect comparison. Updated to `?? false` as per https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md

File: packages/api/src/utils/logs.ts [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    if (message instanceof Error) {
+      log = {
+        ...log,
+        stack: message.stack,
+        message: message.message
+      }
+      if (
+        (this.opts.sentry ?? false) &&
+        (skipForSentry !== undefined) &&
+        !skipForSentry.some((cls) => message instanceof Error)
```
Comment (adamalton) [inline packages/api/src/utils/logs.ts:200]: I'm pretty sure this should be `message instanceof cls`..?

Comment (joshghent) [reply packages/api/src/utils/logs.ts]: Yes correct, this is now sorted.

**Comment by adamalton** [comment]
I've added the `LOGTAIL_TOKEN` to both the staging and production environments for the w3name Cloudflare worker. Just waiting for the `SENTRY_DSN` token now.

**Comment by adamalton** [comment]
Ok, we've now got a GH secret at the organisation level for `SENTRY_TOKEN`. And I've added the `SENTRY_DSN` as an encrypted env variable in the Cloudflare workers(s).

**Comment by mbommerez** [comment]
Left to do:

- [ ] Log in to Sentry, check if staging logs are populated.

**Pull request merged**